contracts/cairo/src/escrow.cairo

```diff
@@ -282,4 +282,43 @@ mod Escrow {
 
         self.emit(Withdraw { order_id, address: self.mm_starknet_wallet.read(), amount });
     }
+
+    #[l1_handler]
+    fn withdraw_batch(
+        ref self: ContractState,
+        from_address: felt252,
+        order_ids: Array<u256>,
+        recipient_addresses: Array<EthAddress>,
+        amounts: Array<u256>
+    ) {
+        let eth_transfer_contract_felt: felt252 = self.eth_transfer_contract.read().into();
+        assert(eth_transfer_contract_felt == from_address, 'Only ETH_TRANSFER_CONTRACT');
+        assert(order_ids.len() == recipient_addresses.len(), 'Different lengths');
+        assert(order_ids.len() == amounts.len(), 'Different lengths');
+        let mut idx = 0;
+
+        loop {
+            if idx >= order_ids.len() {
+                break;
+            }
+
+            let order_id = *order_ids.at(idx);
+            let recipient_address = *recipient_addresses.at(idx);
+            let amount = *amounts.at(idx);
+            assert(!self.orders_used.read(order_id), 'Order withdrew or nonexistent');
+            let order = self.orders.read(order_id);
+            assert(order.recipient_address == recipient_address, 'recipient_address not match L1');
+            assert(order.amount == amount, 'amount not match L1');
```

Comment:
These asserts will revert the whole TX, even if only one order has an error. Is there a way we could maybe skip the orders that have an error? And execute the withdraws that are OK.

This would greatly lower the failure rate of batch withdraws (and save the gas from these failed withdraws!)

Reply:
I think there are some tradeoffs to be had here, it is possible to "skip" failing txs, but it would take more gas for non failing ones, which i think would be the common case. Also, how do you comunicate back to the market maker that some part of the batch failed and some part of it succeed.

The other alternative i thought of was two loops, one with the asserts, to fail fast if one order fails, and the other one that actually executes the orders. Open to discussion on this.

Reply:
As we discussed this morning, "batch withdraw" will be an atomic transaction, so if one fails the whole tx will fail.
I believe this is the way it is currently programmed, so feel free to mark this conversation as resolved 👍